spanner/spannertest/README.md

```diff
@@ -22,7 +22,7 @@ by ascending esotericism:
 - more aggregation functions
 - SELECT HAVING
 - more literal types
-- generated columns
+- generated columns referencing other generated columns
```

Comment:
```suggestion
- generated columns referencing other generated columns or columns containing NULL values
```

It seems that this implementation does not support generating a value based on other columns that contain a NULL value. Is there a specific reason why that cannot be supported by this implementation? If so, we should add a short notice here, but my impression is that there's nothing special here that makes it impossible to generate a value if one of the dependent columns contains a null value.

Reply:
I struggled to find decent documentation but I was playing with live spanner it appeared to not generate any value if the column it depended on was null.

Reply:
My problem was mostly how do I determine when we should calculate a value for a generated column, I guess you are correct that sometimes basing it on a column containing null is in the column, I just didn't want to hide errors, i.e. when trying to multiply a null by 2 I'll get a null and be unable to store it, I wanted to differentiate that from if something else went wrong but it felt non trivial. Perhaps that is better, if the evaluating the expression fails I just store nothing.

Reply:
In that case I don't even care about which columns I'm dependent on as I'm making no promise to resolve generated columns in a sensible order to enable them to depend on each other.

Reply:
The evaluation should normally not fail, but `NULL` should be stored in the generated column. Most expressions that include at least one `NULL` value will evaluate to null, and as long as the generated column is not marked as `NOT NULL` it should not be a problem. Consider the following example:

```
create table Singers (SingerId INT64, FirstName STRING(100), LastName STRING(100), FullName STRING(200) AS (FirstName || ' ' || LastName) STORED) PRIMARY KEY (SingerId);

insert into Singers (SingerId, FirstName, LastName) VALUES (1, 'First', 'Last');
insert into Singers (SingerId, FirstName, LastName) VALUES (2, 'First', NULL);
```

The `Singers` table will now contain two rows, and the `FullName` value for row 2 will be `NULL`, as string concatenation produces `NULL` if one of the concatenated values is null.

Reply:
So in that case I could just run the expression and as long as the type matches we store it, dump all this logic looking for dependent columns, what do you think?

Reply:
Yes, I think that would be better. If the evaluation of the expression returns an error instead of a value, then of course the error should be returned, but otherwise if there is no error then we should just accept the outcome of the evaluation and store it.